rfc/rfc-83/rfc-83.md

```diff
@@ -0,0 +1,260 @@
+
+### Abstraction
+
+Use `TableServiceCommitter` to control table service related commit action.
+
+```java
+package org.apache.hudi.table.action;
+
+public interface TableServiceCommitter<R> {
+
+  /**
+   * Build table service related commit metadata, 
+   * Especially `EarliestCommitToReta` column when enable incremental table service.
+   * @return metadata to commit
+   */
+  R buildCommitMeta();
+}
+```
+
+Use `TableServiceBaseActionExecutor` to control the behavior of getting partitions to be performed into current table service.
+
+```java
+package org.apache.hudi.table.action;
+
+import org.apache.hudi.common.engine.HoodieEngineContext;
+import org.apache.hudi.common.model.TableServiceType;
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.exception.HoodieNotSupportedException;
+import org.apache.hudi.table.HoodieTable;
+
+import java.io.IOException;
+import java.util.List;
+
+public abstract class TableServiceBaseActionExecutor<T, I, K, O, R> extends BaseActionExecutor<T, I, K, O, R> implements TableServiceCommitter<R> {
+
+  public TableServiceBaseActionExecutor(HoodieEngineContext context, HoodieWriteConfig config, HoodieTable<T, I, K, O> table, String instantTime) {
+    super(context, config, table, instantTime);
+  }
+
+  /**
+   * Returns the earliest commit to retain and will record in meta as new EarliestCommitToRetain
+   */
+  public Option<HoodieInstant> getEarliestCommitToRetain() {
+    return null;
+  }
+
+  /**
+   * get partitions to be performed in current table service
+   * @param instantToRetain
+   * @param type
+   * @return
+   * @throws IOException
+   */
+  public List<String> getPartitionPaths(Option<HoodieInstant> instantToRetain, TableServiceType type) {
+    return null;
+  }
+
+  /**
+   * Get incremental partitions from EarliestCommitToRetain to instantToRetain
+   * @param instantToRetain
+   * @param type
+   * @param deleteEmptyCommit
+   * @return
+   * @throws IOException
+   */
+  private List<String> getIncrementalPartitionPaths(Option<HoodieInstant> instantToRetain, TableServiceType type) {
+    return null;
+  }
+
+  /**
+   * Get all table partitions related to current table
+   * @param instantToRetain
+   * @param type
+   * @param deleteEmptyCommit
+   * @return
+   * @throws IOException
+   */
+  private List<String> getAllPartitionPaths(Option<HoodieInstant> instantToRetain, TableServiceType type) {
+    return null;
+  }
+
+  public R buildCommitMeta() {
+    return null;
+  }
+
+}
+
+```
+
+### Optimize the inheritance relationship of Table Service
+
+![NewTableServiceDiagram.png](NewTableServiceDiagram.png)
+
+### Work Flow for Incremental Table Service
+
+Table Service Planner
+1. Retrieve the instant recorded in the last completed table service commit as **INSTANT 1**.
```

Comment:
Do we want to take care of the cases where `INSTANT 1` already been archived?

Reply:
We record `EarliestCommitToRetain` in the TableService Request metadata file and use it as the basis for retrieving incremental partitions. 
Therefore, when Incremental Table Service is enabled, we should always ensure that there is a Clustering/Compaction request metadata in the active timeline.

Also we can use getAllpartitions as Cover-up plan

Reply:
> Therefore, when Incremental Table Service is enabled, we should always ensure that there is a Clustering/Compaction request metadata in the active timeline.

Not really, for cleaning this is true because there could be data quality issues of wrong files are cleaned, but Compaction and Clustering are just rewrites.